Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -2342,19 +2342,22 @@ public virtual async Task OnPipCompleted(RunnablePip runnablePip)
                 // This happens before we traverse the dependents of this failed pip; since SchedulePipIfReady
                 // checks m_scheduleTerminating this means that there will not be 'Skipped' pips in this mode
                 // (they remain unscheduled entirely).
-                if (!succeeded && m_scheduleConfiguration.StopOnFirstError)
+                if (!succeeded && !IsDistributedWorker)
                 {
-                    if (!m_scheduleTerminating)
+                    // We stop on the first error only on the master or single-machine builds.
+                    // During cancellation, master coordinates with workers to stop the build.
+
+                    bool hasMaterializationErrorHappened = m_executePhaseLoggingContext.ErrorsLoggedById.Contains((ushort)EventId.PipMaterializeDependenciesFromCacheFailure)

```

Comment:
>ErrorsLoggedById.Contains [](start = 88, length = 25)

ConcurrentBag.Contains() isn't particularly performant. It copies everything to a new list and then enumerates that. I think it's smart in the empty case and we're only doing this for failed pips so it's probably ok. It might be worth a comment to that effect. Your call

Reply:
added a comment. I would not worry about it as this is only for failed pips and we iterate through the errors.